main.go

```diff
@@ -549,3 +556,88 @@ func uploadFile(uploadURL string, archiveFilePath string) error {
 
 	return nil
 }
+
+func createDimensions(step toolresults.Step) map[string]string {
+	dimensions := map[string]string{}
+	for _, dimension := range step.DimensionValue {
+		dimensions[dimension.Key] = dimension.Value
+	}
+	return dimensions
+}
+
+func printStepsStatesToStartTime(stepsStatesToStartTime map[string]map[string]time.Time, stepsToNames map[string]string, currentTime time.Time, w io.Writer) {
+	var stepIDs []string
+	for stepID := range stepsToNames {
+		stepIDs = append(stepIDs, stepID)
+	}
+
+	sort.Strings(stepIDs)
+
+	for _, stepID := range stepIDs {
+		stepName := stepsToNames[stepID]
+		if _, err := fmt.Fprintln(w, stepName); err != nil {
+			fmt.Printf("Failed to print step status durations: %s", err)
+			return
+		}
+
+		statesToStartTime := stepsStatesToStartTime[stepID]
+
+		var states []string
+		for state := range statesToStartTime {
+			states = append(states, state)
+		}
+
+		sort.Slice(states, func(i, j int) bool {
+			stateI, stateJ := states[i], states[j]
+			startTimeI, startTimeJ := statesToStartTime[stateI], statesToStartTime[stateJ]
+
+			return startTimeI.Before(startTimeJ)
+		})
+
+		for i, state := range states {
+			startTime := statesToStartTime[state]
+
+			var endTime time.Time
+			if i == len(states)-1 {
+				endTime = currentTime
+			} else {
+				nextState := states[i+1]
+				endTime = statesToStartTime[nextState]
+			}
+
+			duration := endTime.Sub(startTime)
+
+			if _, err := fmt.Fprintf(w, "- time spent in %s state: ~%ds\n", state, duration/time.Second); err != nil {
+				fmt.Printf("Failed to print step status durations: %s", err)
+				return
+			}
+		}
+	}
+}
+
+func createStepNameWithDimensions(step toolresults.Step) string {
+	dimensions := createDimensions(step)
+	return fmt.Sprintf("%s (%s %s %s %s)", step.Name, dimensions["Model"], dimensions["Version"], dimensions["Orientation"], dimensions["Locale"])
+}
+
+func updateStepsStatesToStartTime(stepsStatesToStartTime map[string]map[string]time.Time, stepsToNames map[string]string, response toolresults.ListStepsResponse) {
```

Comment:
Why are the step names separated from the states and timings? I think it made it a bit harder to understand for me. 

We could move them into an internal struct and have something like
```
type internalStepData struct {
	name           string
	stateAndTiming map[string]time.Time
}

func newInternalStepData(step toolresults.Step) internalStepData {
	return internalStepData{
		name:           createStepNameWithDimensions(step),
		stateAndTiming: map[string]time.Time{},
	}
}

func updateStepsStatesToStartTime2(steps map[string]internalStepData, response toolresults.ListStepsResponse) {
	for _, step := range response.Steps {
		data, ok := steps[step.StepId]
		if !ok {
			data = newInternalStepData(*step)
			steps[step.StepId] = data
		}

		_, ok = data.stateAndTiming[step.State]
		if !ok {
			// Haven't seen this state yet -> set state start time
			data.stateAndTiming[step.State] = time.Now()
		}
	}
}
```
and then also the time calculation could be moved into this internal struct. This way you might not need to understand so many things upfront. 😄

But it also good as it is. It just my first impression.

Reply:
Updated in https://github.com/bitrise-steplib/steps-virtual-device-testing-for-ios/pull/36/commits/b438236e1e8d2746cc9912772f247c7bf8aca0f2 and https://github.com/bitrise-steplib/steps-virtual-device-testing-for-ios/pull/36/commits/5487a88d8a2a7588bef4f43ff82da1de1e36d392